ext/ddtrace_profiling_native_extension/collectors_discrete_dynamic_sampler.c

```diff
@@ -0,0 +1,351 @@
+  const char *debug_name = sampler->debug_name;
+  double target_overhead = sampler->target_overhead;
+  (*sampler) = (discrete_dynamic_sampler) {
+    .debug_name = debug_name,
+    .target_overhead = target_overhead,
+    // Act as if a reset is a readjustment (it kinda is!) and wait for a full adjustment window
+    // to compute stats. Otherwise, we'd readjust on the next even that comes and thus be operating
+    // with very incomplete information
+    .last_readjust_time_ns = now_ns,
+    // This fake readjustment will use a hardcoded sampling interval
+    .sampling_interval = BASE_SAMPLING_INTERVAL,
+    // But we want to make sure we sample at least once in the next window so that our first
+    // real readjustment has some notion of how heavy sampling is
+    .events_since_last_sample = BASE_SAMPLING_INTERVAL,
+  };
+}
+
+void discrete_dynamic_sampler_reset(discrete_dynamic_sampler *sampler) {
+  long now = monotonic_wall_time_now_ns(DO_NOT_RAISE_ON_FAILURE);
+  _discrete_dynamic_sampler_reset(sampler, now);
+}
+
+static void _discrete_dynamic_sampler_set_overhead_target_percentage(discrete_dynamic_sampler *sampler, double target_overhead, long now_ns) {
+  if (target_overhead <= 0 || target_overhead > 100) {
+    rb_raise(rb_eArgError, "Target overhead must be a double between ]0,100] was %f", target_overhead);
+  }
+  sampler->target_overhead = target_overhead;
+  _discrete_dynamic_sampler_reset(sampler, now_ns);
+}
+
+void discrete_dynamic_sampler_set_overhead_target_percentage(discrete_dynamic_sampler *sampler, double target_overhead) {
+  long now = monotonic_wall_time_now_ns(DO_NOT_RAISE_ON_FAILURE);
+  _discrete_dynamic_sampler_set_overhead_target_percentage(sampler, target_overhead, now);
+}
+
+static void maybe_readjust(discrete_dynamic_sampler *sampler, long now);
+
+static bool _discrete_dynamic_sampler_should_sample(discrete_dynamic_sampler *sampler, long now_ns) {
+  // For efficiency reasons we don't do true random sampling but rather systematic
+  // sampling following a sample interval/skip. This can be biased and hide patterns
+  // but the dynamic interval and rather indeterministic pattern of allocations in
+  // most real applications should help reduce the bias impact.
+  sampler->events_since_last_sample++;
+  sampler->events_since_last_readjustment++;
+  bool should_sample = sampler->sampling_interval > 0 && sampler->events_since_last_sample >= sampler->sampling_interval;
+
+  if (should_sample) {
+    sampler->sample_start_time_ns = now_ns;
+  } else {
+    // check if we should readjust our sampler after this event, even if we didn't sample it
+    maybe_readjust(sampler, now_ns);
+  }
+
+  return should_sample;
+}
+
+bool discrete_dynamic_sampler_should_sample(discrete_dynamic_sampler *sampler) {
+  long now = monotonic_wall_time_now_ns(DO_NOT_RAISE_ON_FAILURE);
+  return _discrete_dynamic_sampler_should_sample(sampler, now);
+}
```

Comment:
I would prefer to raise yes but this would get called outside the rescue wrapper. Maybe we could simply check if we get a 0 return value and trigger a sampler shutdown (e.g. should_sample will always return false and not readjust until it is reset).

Reply:
I think there is enough nuance here to maybe leave it for a different PR? Created https://datadoghq.atlassian.net/browse/PROF-9035. In the meantime, if clocks fail I think that will already trigger profiler stop from the other collectors so unless it can heisen-fail in allocation alone we should be pretty safe.

Reply:
Yup, good point :)